Sources/CoreMetrics/SystemMetrics.swift

```diff
@@ -0,0 +1,172 @@
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of Swift Metrics API project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Combine
+import Foundation
+
+@available(OSX 10.15, *)
+public enum SystemMetricsProvider {
+    fileprivate static let lock = ReadWriteLock()
+    fileprivate static let queue = DispatchQueue(label: "com.apple.CoreMetrics.SystemMetricsProvider", qos: .background)
+    fileprivate static var timeInterval: DispatchTimeInterval = .seconds(2)
+    fileprivate static var cancellable: Cancellable?
+    fileprivate static var systemMetricsType: SystemMetrics.Type = UnsuportedMetricsProvider.self
+
+    public static func bootstrapSystemMetrics(pollInterval interval: DispatchTimeInterval = .seconds(2), systemMetricsType: SystemMetrics.Type? = nil) {
+        self.lock.withWriterLockVoid {
+            self.timeInterval = interval
+            if let type = systemMetricsType {
+                self.systemMetricsType = type
+            } else {
+                #if os(Linux)
+                self.systemMetricsType = LinuxSystemMetricsProvider.self
+                #else
+                self.systemMetricsType = UnsuportedMetricsProvider.self
+                #endif
+            }
+        }
+        self.updateSystemMetrics()
+    }
+
+    public static func cancelSystemMetrics() {
+        self.cancellable?.cancel()
+    }
+
+    fileprivate static func updateSystemMetrics() {
+        let interval = self.lock.withReaderLock { self.timeInterval }
+        let c = self.queue.schedule(after: .init(.now()), interval: .init(interval)) {
+            let prefix = "process_"
+            let pid = ProcessInfo.processInfo.processIdentifier
+            let metrics = self.systemMetricsType.init(pid: "\(pid)")
+            if let vmem = metrics.virtualMemory { Gauge(label: prefix + "virtual_memory_bytes").record(vmem) }
+            if let rss = metrics.residentMemory { Gauge(label: prefix + "resident_memory_bytes").record(rss) }
+            if let start = metrics.startTimeSeconds { Gauge(label: prefix + "start_time_seconds").record(start) }
+            if let cpuSeconds = metrics.cpuSeconds { Gauge(label: prefix + "cpu_seconds_total").record(cpuSeconds) }
+            if let maxFds = metrics.maxFds { Gauge(label: prefix + "max_fds").record(maxFds) }
+            if let openFds = metrics.openFds { Gauge(label: prefix + "open_fds").record(openFds) }
+        }
+        self.lock.withWriterLockVoid {
+            self.cancellable = c
+        }
+    }
+}
+
+public protocol SystemMetrics {
+    init(pid: String)
+
+    var virtualMemory: Int32? { get }
+    var residentMemory: Int32? { get }
+    var startTimeSeconds: Int32? { get }
+    var cpuSeconds: Int32? { get }
+    var maxFds: Int32? { get }
+    var openFds: Int32? { get }
```

Comment:
Are we sure about the `?`? I know it's kind of "right" but all those are numerics where `0` can be used as a default if we could not read... WDYT?

Reply:
The reasoning for optional is the difference between `0` and `nil` for reporting. A metric not being supported and thus not exported is different from that metric being `0`. This could cause issues with (for example) alerting systems. I think in this case it's a better safe than sorry situation. But if I'm alone in this, we can change it 😄

Reply:
- They should all be `Int`, no? why don't we just make them `Int` as quoted above, that's what the [Swift language guide](https://docs.swift.org/swift-book/LanguageGuide/TheBasics.html#ID324) recommends.
- the unit is missing in all the fields. This should be `virtualMemoryBytes`, right? (And then `Int32` is also definitely too small)